server/redeemer.go

```diff
@@ -0,0 +1,398 @@
+		return status.Error(codes.Internal, "context is nil")
+	}
+
+	// Make a channel to receive max float updates
+	//  This check allows to overwrite the channel for testing purposes
+	var maxFloatUpdates chan *net.MaxFloatUpdate
+	maxFloatUpdatesI, ok := r.subs.Load(p.Addr.String())
+	if !ok {
+		maxFloatUpdates = make(chan *net.MaxFloatUpdate)
+		r.subs.Store(p.Addr.String(), maxFloatUpdates)
+		glog.Infof("new MonitorMaxFloat subscriber: %v", p.Addr.String())
+	} else {
+		maxFloatUpdates, ok = maxFloatUpdatesI.(chan *net.MaxFloatUpdate)
+		if !ok {
+			return status.Error(codes.Internal, "maxFloatUpdates is of the wrong type")
+		}
+	}
+
+	// Block so that the stream is over a long-lived connection
+	for {
+		select {
+		case maxFloatUpdate := <-maxFloatUpdates:
+			if err := stream.Send(maxFloatUpdate); err != nil {
+				if err == io.EOF {
+					r.subs.Delete(p.Addr.String())
+					return status.Error(codes.Internal, err.Error())
+				}
+				glog.Errorf("Unable to send maxFloat update to client=%v err=%v", p.Addr.String(), err)
+			}
+		case <-r.quit:
+			return nil
+		case <-stream.Context().Done():
+			return nil
+		}
+	}
+}
+
+func (r *redeemer) MaxFloat(ctx context.Context, req *net.MaxFloatRequest) (*net.MaxFloatUpdate, error) {
+	mf, err := r.sm.MaxFloat(ethcommon.BytesToAddress(req.Sender))
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Errorf("max float error: %v", err).Error())
+	}
+	return &net.MaxFloatUpdate{
+		Sender:   req.Sender,
+		MaxFloat: mf.Bytes(),
+	}, nil
+}
+
+func (r *redeemer) startCleanupLoop() {
+	ticker := time.NewTicker(cleanupLoopTime)
+	for {
+		select {
+		case <-ticker.C:
+			// clean up map entries that haven't been cleared since the last cleanup loop ran
+			r.liveSenders.Range(func(key, value interface{}) bool {
+				if value.(time.Time).Add(cleanupLoopTime).Before(time.Now()) {
+					r.liveSenders.Delete(key)
+				}
+				return true
+			})
+		case <-r.quit:
+			return
+		}
+	}
+}
+
+type redeemerClient struct {
+	rpc      net.TicketRedeemerClient
+	maxFloat map[ethcommon.Address]*big.Int
+	mu       sync.RWMutex
+	quit     chan struct{}
+	sm       pm.SenderManager
+	tm       pm.TimeManager
+}
+
+// NewRedeemerClient instantiates a new client for the ticket redemption service
+// The client implements the pm.SenderMonitor interface
+func NewRedeemerClient(uri *url.URL, sm pm.SenderManager, tm pm.TimeManager) (pm.SenderMonitor, *grpc.ClientConn, error) {
+	conn, err := grpc.Dial(
+		uri.String(),
+		grpc.WithBlock(),
+		grpc.WithTimeout(GRPCConnectTimeout),
+		grpc.WithInsecure(),
+	)
+
+	//TODO: PROVIDE KEEPALIVE SETTINGS
+	if err != nil {
+		glog.Errorf("Did not connect to orch=%v err=%v", uri, err)
+		return nil, nil, fmt.Errorf("Did not connect to orch=%v err=%v", uri, err)
+	}
+	return &redeemerClient{
+		rpc:      net.NewTicketRedeemerClient(conn),
+		sm:       sm,
+		tm:       tm,
+		maxFloat: make(map[ethcommon.Address]*big.Int),
+		quit:     make(chan struct{}),
+	}, conn, nil
```

Comment:
Since we already have `Stop()` method that is used to close the quit channel could we store the ClientConn in the struct and call `Close()` on it instead of having the caller close the ClientConn and call `Stop()`?

Reply:
fixed in 717ef7cb128335ca6b690a98a4a6f908f3ad4cce